File: FitoGrano/CS/WinForm/frAddEditHistoryRoleJobs.cs
// VBConversions Note: VB project level imports
using System.Collections.Generic;
using System;
using System.Linq;
using System.Drawing;
using System.Diagnostics;
using Microsoft.VisualBasic;
using System.Xml.Linq;
using System.Windows.Forms;
// End of VB project level imports

using System.ComponentModel;
using System.Data;
using FitoGrano;

namespace FitoGrano
{
	
	public partial class frAddEditHistoryRoleJobs
	{
		public frAddEditHistoryRoleJobs()
		{
			InitializeComponent();
			
			//Added to support default instance behavour in C#
			if (defaultInstance == null)
				defaultInstance = this;
		}
		
#region Default Instance
		
		private static frAddEditHistoryRoleJobs defaultInstance;
		
		/// <summary>
		/// Added by the VB.Net to C# Converter to support default instance behavour in C#
		/// </summary>
		public static frAddEditHistoryRoleJobs Default
		{
			get
			{
				if (defaultInstance == null)
				{
					defaultInstance = new frAddEditHistoryRoleJobs();
					defaultInstance.FormClosed += new System.Windows.Forms.FormClosedEventHandler(defaultInstance_FormClosed);
				}
				
				return defaultInstance;
			}
			set
			{
				defaultInstance = value;
			}
		}
		
		static void defaultInstance_FormClosed(object sender, System.Windows.Forms.FormClosedEventArgs e)
		{
			defaultInstance = null;
		}
		
#endregion
		internal int AddEdit = 2;
		DataRow iRowCurrent;
		// Таскание формы за любой участок
		public void Control_MouseDown(dynamic sender, System.Windows.Forms.MouseEventArgs e)
		{
			if (e.Button == System.Windows.Forms.MouseButtons.Left)
			{
				this.Cursor = Cursors.NoMove2D;
				sender.Capture = false;
				Windows.Forms.Message temp_msg = Message.Create(this.Handle, 0xA1, new IntPtr(2), IntPtr.Zero);
				this.WndProc(ref temp_msg);
				this.Cursor = Cursors.Default;
			}
		}
		// Закрытие формы
		public void frAddEditHistoryRoleJobs_FormClosing(object sender, FormClosingEventArgs e)
		{
			clsMyUserFunction.RemoveTableDataSet(this);
		}
		// загрузка формы
		public void frAddEditHistoryRoleJobs_Load(object sender, EventArgs e)
		{
			mdVariables.EventChangedControl = false;
			if (AddEdit == 3)
			{
				this.lbNameForm.Text = "Редактировать должности сотрудника";
			}
			this.Location = MousePosition;
			this.txtDtBegin.EditValue = DateTime.Now;
			this.txtDtBegin.Select();
			// дерево должностей
			clsOrganizationStructure RoleJobsFilial = new clsOrganizationStructure() {
					sender = this.cmbRoleJobsFilial.Properties.TreeList,
					TableName = "tlRoleJobsFilial." + this.Name
				};
			this.cmbRoleJobsFilial.Properties.DisplayMember = "RoleJobsString";
			this.cmbRoleJobsFilial.Properties.ValueMember = "Id";
			RoleJobsFilial.GetOrganizationStructure(ExpandedRoot: 1, OnlyCurentFillial: "1");
			tlRoleJobsFilial.NodeCellStyle += RoleJobsFilial.NodeCellStyle;
			// если новая запись выходим из процедуры
			if (AddEdit == 2)
			{
				mdVariables.EventChangedControl = true;
				return;
			}
			// текущая строка в таблице
			iRowCurrent = mdQuery.iDataSet.Tables["HistoryRoleJobs." + frAddEditPerformer.Default.Name].Rows[frAddEditPerformer.Default.gvHistoryRoleJobs.FocusedRowHandle];
			this.cmbRoleJobsFilial.EditValue = iRowCurrent["RoleJobsId"];
			this.txtDtBegin.EditValue = iRowCurrent["DtChange"];
			this.tlRoleJobsFilial.FocusedNode = this.tlRoleJobsFilial.FindNodeByKeyID(iRowCurrent["RoleJobsId"]);
			mdVariables.EventChangedControl = true;
		}
		// отмена закрытия выпадающего дерева
		public void cmbRoleJobsFilial_QueryCloseUp(dynamic sender, CancelEventArgs e)
		{
			if (ReferenceEquals(sender.Properties.TreeList.FocusedNode, null))
			{
				return;
			}
			// выбираем только должность
			e.Cancel = sender.Properties.TreeList.FocusedNode("Level") != 3;
		}
		// событие проверки корректности даты
		public void txtDtBegin_Properties_Validating(dynamic sender, CancelEventArgs e)
		{
			if (CheckingDtBegin())
			{
				sender.ErrorText = null;
				this.btnOk.Enabled = true;
			}
			else
			{
				this.txtDtBegin.Focus();
				this.txtDtBegin.SelectAll();
				this.txtDtBegin.ErrorText = "Указанная дата уже присутствует в истории ...";
				this.btnOk.Enabled = false;
			}
		}
		// событие изменения даты
		public void txtDtBegin_EditValueChanged(object sender, EventArgs e)
		{
			if (mdVariables.EventChangedControl == false)
			{
				return;
			}
			txtDtBegin_Properties_Validating(sender, new CancelEventArgs());
		}
		// Проверка уникальности даты
		private bool CheckingDtBegin()
		{
			DataTable dt = mdQuery.iDataSet.Tables["HistoryRoleJobs." + frAddEditPerformer.Default.Name];
			DateTime dt_NewValue = System.Convert.ToDateTime(this.txtDtBegin.EditValue); // введенная дата
			DateTime dt_dsValue; // дата из таблицы для сравнения
			DateTime dt_dsCurValue; // выбранная дата в таблице
			// если редактирование
			if (AddEdit == 3)
			{
				// сравниваем новую и редактируемую дату
				dt_dsCurValue = System.Convert.ToDateTime(iRowCurrent["DtChange"]);
				if (dt_dsCurValue.ToShortDateString() == dt_NewValue.ToShortDateString())
				{
					return true;
				}
			}
			// проходим по таблице в поисках совпадения
			foreach (DataRow r in dt.Rows)
			{
				if (Information.IsDBNull(r["DtChange"]))
				{
					continue;
				}
				dt_dsValue = System.Convert.ToDateTime(r["DtChange"]);
				// если совпадение найдено
				if (dt_dsValue.ToShortDateString() == dt_NewValue.ToShortDateString())
				{
					return false;
				}
			}
			return true;
		}
		// проверка назначения должности
		private bool CheckingRoleJobs()
		{
			if (clsForSQL.InDataBase(this.cmbRoleJobsFilial.EditValue) == "NULL")
			{
				this.cmbRoleJobsFilial.ErrorText = "Должность не назначена ...";
				this.btnOk.Enabled = false;
				return false;
			}
			else
			{
				this.cmbRoleJobsFilial.ErrorText = null;
				this.btnOk.Enabled = true;
				return true;
			}
		}
		// Ок
		public void btnOk_Click(object sender, EventArgs e)
		{
			if (CheckingRoleJobs() == false)
			{
				return; // проверка назначения должности
			}
			if (ExecuteScalar(
					EXEC PerformersHistoryRJFuction
					@PerformerId == " & frAddEditPerformer.PerformerId & ",
					@DtChange == " & clsForSQL.InDataBase(Me.txtDtBegin.Text) & ",
					@DtChangeOld == " & If(AddEdit = 2, " NULL ", clsForSQL.InDataBase(iRowCurrent(" DtChange "), " DateTime ")) & ",
					@RoleJobsId == " & Me.cmbRoleJobsFilial.EditValue & ",
					@Function == & AddEdit,
					"SetRoleJob"
					) == 1)
					{
					// передача даты новой записи
					frAddEditPerformer.Default.NewRecord_HistoryRoleJobs = DateTime.Parse(this.txtDtBegin.Text);
				this.DialogResult = DialogResult.OK;
			}
		}
		// отмена
		public void btnCancel_Click(object sender, EventArgs e)
		{
			this.DialogResult = DialogResult.Cancel;
		}
		// событие изменения должности
		public void cmbRoleJobsFvilial_EditValueChanged(object sender, EventArgs e)
		{
			CheckingRoleJobs();
		}
	}
}

File: FitoGrano/WinForm/frAddEditHistoryRoleJobs.vb
Imports System.ComponentModel
Imports System.Data

Public Class frAddEditHistoryRoleJobs
    Friend AddEdit As Integer = 2
    Dim iRowCurrent As DataRow
    ' Таскание формы за любой участок
    Private Sub Control_MouseDown(ByVal sender As Object, ByVal e As System.Windows.Forms.MouseEventArgs) Handles PanelControl1.MouseDown, lbNameForm.MouseDown
        If e.Button = Windows.Forms.MouseButtons.Left Then
            Me.Cursor = Cursors.NoMove2D
            sender.Capture = False
            Me.WndProc(Message.Create(Me.Handle, &HA1, New IntPtr(2), IntPtr.Zero))
            Me.Cursor = Cursors.Default
        End If
    End Sub
    ' Закрытие формы
    Private Sub frAddEditHistoryRoleJobs_FormClosing(sender As Object, e As FormClosingEventArgs) Handles Me.FormClosing, btnClose.Click
        clsMyUserFunction.RemoveTableDataSet(Me)
    End Sub
    ' загрузка формы
    Private Sub frAddEditHistoryRoleJobs_Load(sender As Object, e As EventArgs) Handles Me.Load
        EventChangedControl = False
        If AddEdit = 3 Then Me.lbNameForm.Text = "Редактировать должности сотрудника"
        Me.Location = MousePosition
        Me.txtDtBegin.EditValue = Now
        Me.txtDtBegin.Select()
        ' дерево должностей
        Dim RoleJobsFilial As New clsOrganizationStructure With {
                                                    .sender = Me.cmbRoleJobsFilial.Properties.TreeList,
                                                    .TableName = "tlRoleJobsFilial." & Me.Name
                                                                }
        Me.cmbRoleJobsFilial.Properties.DisplayMember = "RoleJobsString"
        Me.cmbRoleJobsFilial.Properties.ValueMember = "Id"
        RoleJobsFilial.GetOrganizationStructure(ExpandedRoot:=1, OnlyCurentFillial:="1")
        AddHandler tlRoleJobsFilial.NodeCellStyle, AddressOf RoleJobsFilial.NodeCellStyle
        ' если новая запись выходим из процедуры
        If AddEdit = 2 Then EventChangedControl = True : Exit Sub
        ' текущая строка в таблице
        iRowCurrent = iDataSet.Tables("HistoryRoleJobs." & frAddEditPerformer.Name).Rows(frAddEditPerformer.gvHistoryRoleJobs.FocusedRowHandle)
        Me.cmbRoleJobsFilial.EditValue = iRowCurrent("RoleJobsId")
        Me.txtDtBegin.EditValue = iRowCurrent("DtChange")
        Me.tlRoleJobsFilial.FocusedNode = Me.tlRoleJobsFilial.FindNodeByKeyID(iRowCurrent("RoleJobsId"))
        EventChangedControl = True
    End Sub
    ' отмена закрытия выпадающего дерева
    Private Sub cmbRoleJobsFilial_QueryCloseUp(sender As Object, e As CancelEventArgs) Handles cmbRoleJobsFilial.QueryCloseUp
        If sender.Properties.TreeList.FocusedNode Is Nothing Then Exit Sub
        ' выбираем только должность
        e.Cancel = sender.Properties.TreeList.FocusedNode("Level") <> 3
    End Sub
    ' событие проверки корректности даты
    Private Sub txtDtBegin_Properties_Validating(sender As Object, e As CancelEventArgs) Handles txtDtBegin.Properties.Validating
        If CheckingDtBegin() Then
            sender.ErrorText = Nothing
            Me.btnOk.Enabled = True
        Else
            Me.txtDtBegin.Focus()
            Me.txtDtBegin.SelectAll()
            Me.txtDtBegin.ErrorText = "Указанная дата уже присутствует в истории ..."
            Me.btnOk.Enabled = False
        End If
    End Sub
    ' событие изменения даты
    Private Sub txtDtBegin_EditValueChanged(sender As Object, e As EventArgs) Handles txtDtBegin.EditValueChanged
        If EventChangedControl = False Then Exit Sub
        txtDtBegin_Properties_Validating(sender, New CancelEventArgs)
    End Sub
    ' Проверка уникальности даты
    Private Function CheckingDtBegin() As Boolean
        Dim dt As DataTable = iDataSet.Tables("HistoryRoleJobs." & frAddEditPerformer.Name)
        Dim dt_NewValue As Date = Me.txtDtBegin.EditValue   ' введенная дата 
        Dim dt_dsValue As Date                              ' дата из таблицы для сравнения
        Dim dt_dsCurValue As Date                           ' выбранная дата в таблице
        ' если редактирование
        If AddEdit = 3 Then
            ' сравниваем новую и редактируемую дату
            dt_dsCurValue = iRowCurrent("DtChange")
            If dt_dsCurValue.ToShortDateString = dt_NewValue.ToShortDateString Then Return True
        End If
        ' проходим по таблице в поисках совпадения
        For Each r As DataRow In dt.Rows
            If IsDBNull(r("DtChange")) Then Continue For
            dt_dsValue = r("DtChange")
            ' если совпадение найдено
            If dt_dsValue.ToShortDateString = dt_NewValue.ToShortDateString Then
                Return False
            End If
        Next
        Return True
    End Function
    ' проверка назначения должности
    Private Function CheckingRoleJobs() As Boolean
        If clsForSQL.InDataBase(Me.cmbRoleJobsFilial.EditValue) = "NULL" Then
            Me.cmbRoleJobsFilial.ErrorText = "Должность не назначена ..."
            Me.btnOk.Enabled = False
            Return False
        Else
            Me.cmbRoleJobsFilial.ErrorText = Nothing
            Me.btnOk.Enabled = True
            Return True
        End If
    End Function
    ' Ок
    Private Sub btnOk_Click(sender As Object, e As EventArgs) Handles btnOk.Click
        If CheckingRoleJobs() = False Then Exit Sub ' проверка назначения должности  
        If ExecuteScalar(
                        "EXEC PerformersHistoryRJFuction 
                            @PerformerId = " & frAddEditPerformer.PerformerId & ", 
                            @DtChange = " & clsForSQL.InDataBase(Me.txtDtBegin.Text) & ", 
                            @DtChangeOld = " & If(AddEdit = 2, "NULL", clsForSQL.InDataBase(iRowCurrent("DtChange"), "DateTime")) & ", 
                            @RoleJobsId = " & Me.cmbRoleJobsFilial.EditValue & ", 
                            @Function = " & AddEdit,
                        "SetRoleJob"
                              ) = 1 Then
            ' передача даты новой записи
            frAddEditPerformer.NewRecord_HistoryRoleJobs = Me.txtDtBegin.Text
            Me.DialogResult = DialogResult.OK
        End If
    End Sub
    ' отмена
    Private Sub btnCancel_Click(sender As Object, e As EventArgs) Handles btnCancel.Click
        Me.DialogResult = DialogResult.Cancel
    End Sub
    ' событие изменения должности
    Private Sub cmbRoleJobsFvilial_EditValueChanged(sender As Object, e As EventArgs) Handles cmbRoleJobsFilial.EditValueChanged
        CheckingRoleJobs()
    End Sub
End Class